### PR TITLE
Add Macronix Flash support for intel_s1000_crb board.

### PR DESCRIPTION
**Pull request by SavinayDharmappa**
Below are contents of the patch set,

1. Enable SPI Master driver for intel_s1000

2.  Add Macronix Flash support  to generic SPI nor flash drive proposed in following PR #9078 (https://github.com/zephyrproject-rtos/zephyr/pull/9078)

3. Add test application to read/write from macronix flash

### CODE REVIEW

**Review by vanwinkeljan — COMMENTED** [review]
Please take in to consideration following comment that was placed on #9078:
> Looks to me that there is already more generic SPI NOR driver pull request #7561 (which is based on pr #5456). PR #7561 relies on the SFDP information present in the flash device instead of adding the data hard coded into a struct for every device.
>
>Is till see a use case for this driver if the parameters where passed from dts or kconfig as this would reduce the code size compared to auto detection at run time.

File: tests/boards/intel_s1000_crb/src/main.c [hunk 1]
```diff
@@ -19,4 +19,7 @@ K_SEM_DEFINE(thread_sem, 1, 1);
 void main(void)
 {
 	printk("Sample app running on: %s Intel S1000 CRB\n", CONFIG_ARCH);
+	printk("Sample app running on: %s Intel_S1000\n", CONFIG_ARCH);
```
Comment (rgundi) [inline tests/boards/intel_s1000_crb/src/main.c]: Retain the old printk please.

File: tests/boards/intel_s1000_crb/src/spi_flash.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	flash_dev = device_get_binding(CONFIG_SPI_NOR_DRV_NAME);
+
+	if (!flash_dev) {
+		printk("SPI flash driver was not found!\n");
+		return;
+	}
+
+	printk("SPI flash driver was found!\n");
+
+	printk("\nTest 1: Flash erase\n");
```
Comment (rgundi) [inline tests/boards/intel_s1000_crb/src/spi_flash.c]: Incorrect comment. Actually, no comment is required for the below operation.

File: tests/boards/intel_s1000_crb/src/spi_flash.c [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	flash_write_protection_set(flash_dev, false);
+
+	printk("\nTest 1: Flash erase\n");
+
+
```
Comment (rgundi) [inline tests/boards/intel_s1000_crb/src/spi_flash.c]: remove the empty line.

File: tests/boards/intel_s1000_crb/src/spi_flash.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	if (flash_erase(flash_dev,
+			FLASH_TEST_REGION_OFFSET,
+			FLASH_SECTOR_SIZE) != 0) {
+		printk("   Flash erase failed!\n");
+	} else {
+		printk("   Flash erase succeeded!\n");
+	}
+
+
```
Comment (rgundi) [inline tests/boards/intel_s1000_crb/src/spi_flash.c]: remove the empty line.

File: tests/boards/intel_s1000_crb/src/spi_flash.c [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	printf("\nTest 2: Flash write\n");
+	flash_write_protection_set(flash_dev, false);
+
+	buf[0] = TEST_DATA_BYTE_0;
+	buf[1] = TEST_DATA_BYTE_1;
+
+	for (i = 2 ; i < TEST_DATA_LEN; i++) {
+		buf[i] = 0x63;
+	}
```
Comment (rgundi) [inline tests/boards/intel_s1000_crb/src/spi_flash.c]: not sure the above for loop is intended. It rewrites buf[0] and buf[1] with 0x63. Probably you need to delete the for loop.

File: tests/boards/intel_s1000_crb/src/spi_flash.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#define FLASH_TEST_REGION_OFFSET 0x10000
+#define FLASH_SECTOR_SIZE        0x10000
+#define TEST_DATA_BYTE_0         0x4f
+#define TEST_DATA_BYTE_1         0x4a
+#define TEST_DATA_LEN            128
+
+void test_flash(void)
+{
+	struct device *flash_dev;
+	u8_t buf[TEST_DATA_LEN + 10];
```
Comment (rgundi) [inline tests/boards/intel_s1000_crb/src/spi_flash.c]: any particular reason why buf is TEST_DATA_LEN +10? Please remove +10, if it is not needed.

File: tests/boards/intel_s1000_crb/src/spi_flash.c [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+			FLASH_TEST_REGION_OFFSET,
+			FLASH_SECTOR_SIZE) != 0) {
+		printk("   Flash erase failed!\n");
+	} else {
+		printk("   Flash erase succeeded!\n");
+	}
+
+
+	printf("\nTest 2: Flash write\n");
+	flash_write_protection_set(flash_dev, false);
```
Comment (rgundi) [inline tests/boards/intel_s1000_crb/src/spi_flash.c]: since we are already doing this on line 36, do we need to do it again?

**Review by rgundi — CHANGES_REQUESTED** [review]
@vanwinkeljan: Agree PR #7561 is more generic. However, I wanted to know if ALL Flash devices have SFDP information present in them. If not, i believe there's place for this driver to address such cases.

Of course, I agree that the ID parameters can be passed through KConfig or DTS to make it more configurable. I think that should be done if we agree to accept this driver.

**Comment by vanwinkeljan** [comment]
@rgundi 
Just a side note even if all flash devices would support SFDP, which is possibly not the case, there is still a case for this driver. E.g in case you are making your own HW the chance is very small that you will have different flash devices in a single HW version and as such you do not need the overhead of the SFDP handling.

**Comment by SavinayDharmappa** [comment]
@vanwinkeljan I have added support to get ID parameters from Kconfig. Can you review it.

File: drivers/flash/Kconfig.nor [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	prompt "SPI flash device name"
+	default "SPI_NOR"
+	help
+	  This option is mandatory to identify the NOR chip and retrieve its
+	  characterics.
+	  it actually support the following NOR chips:
+	   - W25Q80
+	   - W25Q16
+	   - W25Q32
+	   - S25FL216K
```
Comment (rgundi) [inline drivers/flash/Kconfig.nor]: Add MX25UM512

File: soc/xtensa/intel_s1000/soc.h [hunk 9]
```diff
@@ -154,6 +154,7 @@ struct soc_dsp_shim_regs {
 #define USB_DW_BASE				0x000A0000
 #define USB_DW_IRQ				0x00000806
 
+#define SPI_DW_IRQ_FLAGS                        0
```
Comment (rgundi) [inline soc/xtensa/intel_s1000/soc.h]: Is this required? If yes, i thought we get it from DTS.

**Review by rgundi — CHANGES_REQUESTED** [review]
Looks almost good. Some minor comments.
1) The commit message for "intel_s1000: Enable SPI Master driver" has a typo. Please correct intel_s100 to intel_s1000.
2) The commit "board: xtensa: intel_s1000_crb: Add macronix flash support" needs to come before the commit "tests: boards: intel_s1000: Add a test app to read/write from flash".

File: drivers/flash/spi_nor.c [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+	u32_t sector_size;
+	u32_t n_sectors;
+	u32_t block_size;
+	u32_t flag;
+} flash_ids[] = {
+	{ CONFIG_SPI_NOR_DRV_NAME, JEDEC_ID(CONFIG_SPI_NOR_JEDEC_ID),
+	CONFIG_SPI_NOR_PAGE_SIZE, CONFIG_SPI_NOR_SECTOR_SIZE,
+	CONFIG_SPI_NOR_SECTORS, CONFIG_SPI_NOR_BLOCK_SIZE,
+	CONFIG_SPI_NOR_BLOCK_ERASE_64K},
+};
```
Comment (vanwinkeljan) [inline drivers/flash/spi_nor.c]: All this information can passed as a device configuration struct that to ```DEVICE_AND_API_INIT```, in this case there is no need anymore for ```char* name```.

File: drivers/flash/spi_nor.c [hunk 11]
```diff
@@ -0,0 +1,522 @@
...
+
+/**
+ * struct spi_nor_parameters - Structure for defining the SPI NOR parameters
+ * @flash_size:	Size of the flash
+ * @sector_size: Size of the sectors
+ * @block_size:	Size of the blocks
+ * @page_size: Size of the pages
+ * @flag: Flags of the flash
+ */
+struct spi_nor_parameters {
```
Comment (vanwinkeljan) [inline drivers/flash/spi_nor.c]: In case a device configuration is used this struct becomes obsolete.

**Review by vanwinkeljan — CHANGES_REQUESTED** [review]

**Comment by SavinayDharmappa** [comment]
@vanwinkeljan I have addressed your comments. Please review it.

**Comment by SavinayDharmappa** [comment]
@rgundi  Odering of coommit is changing in git hub only. In my local it is proper. Not sure what is the issue

File: boards/xtensa/intel_s1000_crb/Kconfig.defconfig [hunk 12]
```diff
@@ -63,6 +63,12 @@ config HEAP_MEM_POOL_SIZE
 
 endif # DMA_CAVS
 
+config SPI_0_IRQ_PRI
+	default 3
```
Comment (tbursztyka) [inline boards/xtensa/intel_s1000_crb/Kconfig.defconfig]: you don't need that here, put it in the relevant fixup file (as you for other settings below), and s1000 has no irq priority support afaik (can't recall) so it would be 0.

File: tests/boards/intel_s1000_crb/src/spi_flash.c [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+#define TEST_DATA_BYTE_1         0x4a
+#define TEST_DATA_LEN            128
+
+void test_flash(void)
+{
+	struct device *flash_dev;
+	u8_t buf[TEST_DATA_LEN];
+	int i;
+
+	printf("==========================\n");
```
Comment (tbursztyka) [inline tests/boards/intel_s1000_crb/src/spi_flash.c]: s/printf/printk

But you could also use the logging subsystem which is nicer (since you have macros for info/error/warning and debug)

File: dts/bindings/slave/spi,flash.yaml [hunk 14]
```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2018, Intel Semiconductor
```
Comment (tbursztyka) [inline dts/bindings/slave/spi,flash.yaml]: Intel Corporation

File: dts/bindings/slave/spi,flash.yaml [hunk 15]
```diff
@@ -0,0 +1,25 @@
+#
```
Comment (tbursztyka) [inline dts/bindings/slave/spi,flash.yaml]: Make a patch just for adding that yaml binding file.

File: drivers/flash/Kconfig.nor [hunk 16]
```diff
@@ -27,11 +27,6 @@ config SPI_NOR_DRV_NAME
 	   - S25FL216K
 	   - MX25UM512
 
-config SPI_NOR_SPI_NAME
```
Comment (tbursztyka) [inline drivers/flash/Kconfig.nor]: Why this removal? (you use that option in your board's kconfig)

File: boards/xtensa/intel_s1000_crb/Kconfig.defconfig [hunk 17]
```diff
@@ -69,6 +69,37 @@ config SPI_0_IRQ_PRI
...
 	default 32
 
+config SPI_NOR
+	def_bool y
+
+config SPI_NOR_SPI_SLAVE
+	default 0
+
+config SPI_NOR_DRV_NAME
+	default "MX25"
```
Comment (tbursztyka) [inline boards/xtensa/intel_s1000_crb/Kconfig.defconfig]: that name could come from DTS already, as well as the SPI frequency below. (use fixup to relate these.)

**Review by tbursztyka — DISMISSED** [review]
DTS integration needs more work

File: drivers/flash/Kconfig.nor (hunk 16 above)
Comment (vanwinkeljan) [reply drivers/flash/Kconfig.nor]: @tbursztyka because all SPI related info should come from DTS (see #10894)

Comment (tbursztyka) [reply drivers/flash/Kconfig.nor]: Sure (see my comment about Kconfig.defconfig, it's using Kconfig option while it should get these from DTS), but that change should not be in this patch. It should come in a different one.

Comment (vanwinkeljan) [reply drivers/flash/Kconfig.nor]: Actually I would assume it is part of the first commit as it is introducing a new generic SPI driver

Comment (tbursztyka) [reply drivers/flash/Kconfig.nor]: True

File: tests/boards/intel_s1000_crb/src/spi_flash.c (hunk 13 above)
Comment (SavinayDharmappa) [reply tests/boards/intel_s1000_crb/src/spi_flash.c]: done.

**Comment by SavinayDharmappa** [comment]
@tbursztyka I have adrresssed your review comments . Please review it.

File: soc/xtensa/intel_s1000/dts_fixup.h [hunk 18]
```diff
@@ -46,4 +46,17 @@
 
 #define CONFIG_I2C_0_IRQ_FLAGS			SNPS_DESIGNWARE_I2C_80400_IRQ_0_SENSE
 #define CONFIG_I2C_0_IRQ_PRI			SNPS_DESIGNWARE_I2C_80400_IRQ_0_PRIORITY
+
+
+#define CONFIG_SPI_0_BASE_ADDRESS		SNPS_DESIGNWARE_SPI_E000_BASE_ADDRESS
+#define CONFIG_SPI_0_NAME			SNPS_DESIGNWARE_SPI_E000_LABEL
+
+#define CONFIG_SPI_0_IRQ			((SNPS_DESIGNWARE_SPI_E000_IRQ_0 << 16) | \
```
Comment (tbursztyka) [inline soc/xtensa/intel_s1000/dts_fixup.h]: why this complex macro? Can't you just set the right irq number directly in the .dtsi file?

File: include/arch/xtensa/sys_io.h [hunk 19]
```diff
@@ -115,6 +115,10 @@ static ALWAYS_INLINE
 	return ret;
 }
 
+static ALWAYS_INLINE void sys_write16(u16_t data, mem_addr_t addr)
```
Comment (tbursztyka) [inline include/arch/xtensa/sys_io.h]: Make it in a separate patch.

Also, move it between sys_write/read32 and sys_write/read8, and add sys_read16 as well.

File: tests/boards/intel_s1000_crb/src/main.c [hunk 20]
```diff
@@ -19,4 +22,6 @@ K_SEM_DEFINE(thread_sem, 1, 1);
 void main(void)
 {
 	printk("Sample app running on: %s Intel S1000 CRB\n", CONFIG_ARCH);
```
Comment (tbursztyka) [inline tests/boards/intel_s1000_crb/src/main.c]: LOG_INF?

File: boards/xtensa/intel_s1000_crb/Kconfig.defconfig [hunk 21]
```diff
@@ -66,6 +66,34 @@ endif # DMA_CAVS
 config SPI_DW_FIFO_DEPTH
 	default 32
 
+config SPI_NOR
+	def_bool y
+
+config SPI_NOR_SPI_SLAVE
+	default 0
+
+config SPI_NOR_SPI_FREQ_0
```
Comment (tbursztyka) [inline boards/xtensa/intel_s1000_crb/Kconfig.defconfig]: that one you can get it also from dts.

File: boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts [hunk 22]
```diff
@@ -31,3 +31,14 @@
...
 	};
 };
+
+&spi0 {
+	status = "ok";
+
+	mx25um512@0 {
+		compatible = "spi,flash";
+		label = "MX25UM512";
+		reg = <0x18>;
```
Comment (tbursztyka) [inline boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts]: why 0x18?

File: drivers/flash/Kconfig.nor [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+	help
+	  This option is mandatory to identify the NOR chip and retrieve its
+	  characterics.
+	  it actually support the following NOR chips:
+	   - W25Q80
+	   - W25Q16
+	   - W25Q32
+	   - S25FL216K
+
+config SPI_NOR_SPI_NAME
```
[the remark this answers is not in the document]
Comment (vanwinkeljan) [reply drivers/flash/Kconfig.nor]: Kconfig still contains SPI interface specific config that should move to DTS:
* SPI_NOR_SPI_FREQ_0
* SPI_NOR_SPI_SLAVE
* SPI_NOR_GPIO_SPI_CS
* SPI_NOR_GPIO_SPI_CS_DRV_NAME
* SPI_NOR_GPIO_SPI_CS_PIN

File: soc/xtensa/intel_s1000/dts_fixup.h (hunk 18 above)
Comment (SavinayDharmappa) [reply soc/xtensa/intel_s1000/dts_fixup.h]: No it is not possible to get single interrupt number from .dtsi.

File: boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts (hunk 22 above)
Comment (SavinayDharmappa) [reply boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts]: changed it to 0.

File: boards/xtensa/intel_s1000_crb/Kconfig.defconfig (hunk 21 above)
Comment (SavinayDharmappa) [reply boards/xtensa/intel_s1000_crb/Kconfig.defconfig]: done

File: tests/boards/intel_s1000_crb/src/main.c (hunk 20 above)
Comment (SavinayDharmappa) [reply tests/boards/intel_s1000_crb/src/main.c]: done

File: include/arch/xtensa/sys_io.h (hunk 19 above)
Comment (SavinayDharmappa) [reply include/arch/xtensa/sys_io.h]: ok

File: drivers/flash/Kconfig.nor [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+	default n
+	help
+	This Enables 32K Block erase
+
+config SPI_NOR_BLOCK_ERASE_64K
+	bool "Select 64K block erase"
+	default n
+	help
+	This Enables 64K Block erase
+
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply drivers/flash/Kconfig.nor]: I have done changes and updateded

File: drivers/flash/Kconfig.nor [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+config SPI_NOR_BLOCK_ERASE_64K
+	bool "Select 64K block erase"
+	help
+	This Enables 64K Block erase
+
+endchoice
+
+config SPI_NOR_BLOCK_ERASE_SIZE
+	int "Block size for erasing SPI flash"
```
Comment (vanwinkeljan) [inline drivers/flash/Kconfig.nor]: Just a remark in case you would like that the user can only select values through the choice above remove the text after ```int``` and it will not show up in menu config

File: drivers/flash/Kconfig.nor (hunk 23 above)
[the remark this answers is not in the document]
Comment (vanwinkeljan) [reply drivers/flash/Kconfig.nor]: I think you can already safely remove the GPIO related configs from the KConfig and only keep it in the source file. In this case once you the update for arduino_101 you only need top update the dts and fixup files.

File: drivers/flash/Kconfig.nor (hunk 25 above)
Comment (SavinayDharmappa) [reply drivers/flash/Kconfig.nor]: done.

File: drivers/flash/Kconfig.nor [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+	depends on SPI_NOR_BLOCK_ERASE_64K || SPI_NOR_BLOCK_ERASE_32K
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply drivers/flash/Kconfig.nor]: I have done changes and updated

**Review by vanwinkeljan — APPROVED** [review]

**Review by rgundi — APPROVED** [review]

**Comment by vanwinkeljan** [comment]
@tbursztyka Should we not just close #PR 9078 as this PR  already includes the requested improvements?

File: drivers/flash/spi_nor.c [hunk 27]
```diff
@@ -0,0 +1,453 @@
...
+
+	if (is_write) {
+		return spi_write(driver_data->spi,
+			&driver_data->spi_cfg, &tx_set);
+	}
+
+	return spi_transceive(driver_data->spi,
+		&driver_data->spi_cfg, &tx_set, &rx_set);
+}
+#define spi_nor_cmd_read(dev, opcode, dest, length) \
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: add an empty line right before

File: drivers/flash/spi_nor.c [hunk 28]
```diff
@@ -0,0 +1,453 @@
...
+	return 0;
+}
+
+/**
+ * @brief Wait until the flash is ready
+ *
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int spi_nor_wait_until_ready(struct device *dev)
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: no need to inline

File: drivers/flash/spi_nor.c [hunk 29]
```diff
@@ -0,0 +1,453 @@
...
+	u8_t reg;
+
+	do {
+		ret = spi_nor_cmd_read(dev, SPI_NOR_CMD_RDSR, &reg, 1);
+	} while (!ret && (reg & SPI_NOR_WIP_BIT));
+
+	return ret;
+}
+
+/**
```
Comment (tbursztyka) [inline drivers/flash/spi_nor.c]: remove all flash API documentation: it's redundant with the generic flash header.

**Comment by tbursztyka** [comment]
@vanwinkeljan sure, I added comments then on the first commit that I did on the original PR.

**Comment by rgundi** [comment]
@tbursztyka Could you please check and approve?

File: boards/xtensa/intel_s1000_crb/intel_s1000_crb_defconfig [hunk 30]
```diff
@@ -52,3 +52,8 @@ CONFIG_UART_NS16550=y
 CONFIG_AUDIO=y
 CONFIG_AUDIO_CODEC=y
 CONFIG_AUDIO_TLV320DAC=y
+
+CONFIG_SPI=y
```
Comment (nashif) [inline boards/xtensa/intel_s1000_crb/intel_s1000_crb_defconfig]: Do not enable those by default in the _defconfig, they should be set by the application.

https://docs.zephyrproject.org/latest/porting/board_porting.html#default-board-configuration

**Review by nashif — CHANGES_REQUESTED** [review]

File: boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts [hunk 31]
```diff
@@ -31,3 +31,14 @@
...
 };
+
+&spi0 {
+	status = "ok";
+
+	mx25um512@0 {
+		compatible = "spi,flash";
+		label = "MX25UM512";
+		reg = <0x00>;
+		spi-max-frequency = <2000000>;
```
Comment (sathishkuttan) [inline boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts]: Since this is max frequency, it should be 100MHz. The actual operating frequency (if there is a variable for that) can be 2MHz.

Comment (SavinayDharmappa) [reply boards/xtensa/intel_s1000_crb/intel_s1000_crb.dts]: I don't think we have any such variable in dts.

**Comment by nashif** [comment]
rebase still needed

File: boards/xtensa/intel_s1000_crb/intel_s1000_crb_defconfig (hunk 30 above)
Comment (SavinayDharmappa) [reply boards/xtensa/intel_s1000_crb/intel_s1000_crb_defconfig]: done.

**Comment by SavinayDharmappa** [comment]
re-based and resolved merge conflict.

**Review by nashif — APPROVED** [review]
